sources/packages/backend/libs/services/src/notifications/notification/notification.service.ts

```diff
@@ -70,57 +80,130 @@ export class NotificationService extends RecordDataModelService<Notification> {
+   * @param notificationsSuccessfullyProcessed optional param used to
+   * increment count on recursion.
+   * @returns processing summary.
    */
-  async sendEmailNotification(
-    notificationId: number,
-    entityManager?: EntityManager,
-  ): Promise<GCNotifyResult> {
-    const notificationRepo =
-      entityManager?.getRepository(Notification) ?? this.repo;
-    const notification = await notificationRepo.findOne({
-      select: {
-        id: true,
-        messagePayload: true,
-      },
-      where: {
-        id: notificationId,
-      },
-    });
-    // Call GC Notify send email method.
-    const gcNotifyResult = await this.gcNotifyService.sendEmailNotification(
-      notification.messagePayload,
-    );
+  async processUnsentNotifications(
+    pollingLimit: number,
+    notificationsProcessed = 0,
+    notificationsSuccessfullyProcessed = 0,
+  ): Promise<NotificationProcessingStatus> {
+    const notificationsToProcess = await this.repo
+      .createQueryBuilder("notification")
+      .select(["notification.id", "notification.messagePayload"])
+      .where("notification.dateSent IS NULL")
+      .limit(pollingLimit)
+      .getMany();
+
+    if (notificationsToProcess.length) {
+      this.logger.log(
+        `Processing ${notificationsToProcess.length} notifications`,
+      );
+
+      // Used to limit the number of asynchronous operations
+      // that will start at the same time.
+      const maxPromisesAllowed = os.cpus().length;
+
+      // Hold all the promises that must be processed.
+      const promises: Promise<boolean>[] = [];
+
+      for (const notification of notificationsToProcess) {
+        promises.push(this.sendEmailNotification(notification));
 
-    // Update date sent column in notification table after sending email notification successfully.
-    await this.updateNotification(notification.id, entityManager);
+        if (promises.length >= maxPromisesAllowed) {
+          // Waits for all be executed.
+          const processingResult = await Promise.all(promises);
+          notificationsSuccessfullyProcessed =
+            notificationsSuccessfullyProcessed +
+            processingResult.filter((result) => result).length;
+          // Clear the array.
+          promises.splice(0, promises.length);
+        }
+      }
 
-    return gcNotifyResult;
+      if (promises.length > 0) {
+        // Waits for methods, if any outside the loop.
+        const processingResult = await Promise.all(promises);
+        notificationsSuccessfullyProcessed =
+          notificationsSuccessfullyProcessed +
+          processingResult.filter((result) => result).length;
+      }
+      //Assign the value for total notifications processed.
+      notificationsProcessed =
+        notificationsProcessed + notificationsToProcess.length;
+
+      // Calling process notification in recursion until all the notifications
+      // are processed.
+      const response = await this.processUnsentNotifications(
```

Comment:
I am wondering, so will there be an overlap between 2 consecutive notification queue jobs

Reply:
do we really need this recursion? anyway, jobs are running every minute?

Reply:
To the first question - The scheduler jobs in bull does not overlap, if job 1 takes more than one minute job 2 will not start until job 1 is finished.

to second question - As per our sync up discussion, most of the times 100 records if present is expected to be finished in few seconds and we thought it would be a good idea not to wait until the next minute to pick the next 100 and process recursively in same job.

Reply:
@ann-aot wonder that the GC notify was out of service for 24h and we have 1000 messages to be sent.
With the recursion, we can send the 1000 messages in a few seconds, without the recursion it will be 10min, right assuming a 100 pooling.